File: crates/sofar-core/src/update_cache.rs
//! The update cache (`cli/update-check.ts`): `$XDG_STATE_HOME/sofar/update.json`
//! read for the statusline's segment. The refresh claim-and-spawn stays with
//! the TypeScript surfaces (rust-core O2 ruling); this only ever reads.

use std::path::PathBuf;

use crate::json::{self, Json};
use crate::text::js_trim;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct UpdateCache {
    pub latest: Option<String>,
    pub checked_at: String,
    pub installed: Option<(String, String)>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct UpdateNotice {
    pub latest: String,
    pub current: String,
    pub installed: bool,
}

/// `updateCachePath`.
#[must_use]
pub fn update_cache_path() -> PathBuf {
    let base = std::env::var_os("XDG_STATE_HOME")
        .filter(|v| !js_trim(&v.to_string_lossy()).is_empty())
        .map_or_else(
            || {
                std::env::var_os("HOME")
                    .map_or_else(|| PathBuf::from("/"), PathBuf::from)
                    .join(".local")
                    .join("state")
            },
            PathBuf::from,
        );
    base.join("sofar").join("update.json")
}

/// `readUpdateCache`: the cache, or None when absent/unreadable/corrupt.
#[must_use]
pub fn read_update_cache() -> Option<UpdateCache> {
    let bytes = std::fs::read(update_cache_path()).ok()?;
    let Json::Obj(o) = json::parse(&String::from_utf8_lossy(&bytes)).ok()? else {
        return None;
    };
    let checked_at = o.get("checked_at")?.as_nonempty_str()?.to_owned();
    let latest = o
        .get("latest")
        .and_then(Json::as_nonempty_str)
        .map(str::to_owned);
    let installed = o.get("installed").and_then(Json::as_obj).and_then(|i| {
        Some((
            i.get("version")?.as_str()?.to_owned(),
            i.get("at")?.as_str()?.to_owned(),
        ))
    });
    Some(UpdateCache {
        latest,
        checked_at,
        installed,
    })
}

struct Version {
    release: [u64; 3],
    pre: Option<Vec<String>>,
}

/// `parseVersion`: `^v?(\d+)\.(\d+)\.(\d+)(?:-([0-9A-Za-z.-]+))?(?:\+[0-9A-Za-z.-]+)?$` on the trimmed value.
fn parse_version(value: &str) -> Option<Version> {
    let s = js_trim(value);
    let s = s.strip_prefix('v').unwrap_or(s);
    let ident = |c: char| c.is_ascii_alphanumeric() || c == '.' || c == '-';
    let (core, rest) = match s.find(['-', '+']) {
        Some(i) => (&s[..i], &s[i..]),
        None => (s, ""),
    };
    let mut parts = core.split('.');
    let mut release = [0u64; 3];
    for slot in &mut release {
        let p = parts.next()?;
        if p.is_empty() || !p.bytes().all(|b| b.is_ascii_digit()) {
            return None;
        }
        *slot = p.parse().ok()?;
    }
    if parts.next().is_some() {
        return None;
    }
    let (pre, build) = if let Some(r) = rest.strip_prefix('-') {
        match r.find('+') {
            Some(i) => (Some(&r[..i]), &r[i..]),
            None => (Some(r), ""),
        }
    } else {
        (None, rest)
    };
    if let Some(p) = pre
        && (p.is_empty() || !p.chars().all(ident))
    {
        return None;
    }
    if let Some(b) = build.strip_prefix('+')
        && (b.is_empty() || !b.chars().all(ident))
    {
        return None;
    } else if !build.is_empty() && !build.starts_with('+') {
        return None;
    }
    Some(Version {
        release,
        pre: pre.map(|p| p.split('.').map(str::to_owned).collect()),
    })
}

#[allow(clippy::float_cmp, reason = "`Number(x) !== Number(y)`, verbatim")]
fn compare_prerelease(a: &[String], b: &[String]) -> std::cmp::Ordering {
    use std::cmp::Ordering;
    for i in 0..a.len().max(b.len()) {
        let (Some(x), Some(y)) = (a.get(i), b.get(i)) else {
            return if a.get(i).is_none() {
                Ordering::Less
            } else {
                Ordering::Greater
            };
        };
        let xn = !x.is_empty() && x.bytes().all(|c| c.is_ascii_digit());
        let yn = !y.is_empty() && y.bytes().all(|c| c.is_ascii_digit());
        if xn && yn {
            let (xv, yv): (f64, f64) = (x.parse().unwrap_or(0.0), y.parse().unwrap_or(0.0));
            if xv != yv {
                return if xv < yv {
                    Ordering::Less
                } else {
                    Ordering::Greater
                };
            }
        } else if xn != yn {
            return if xn {
                Ordering::Less
            } else {
                Ordering::Greater
            };
        } else if x != y {
            return crate::text::cmp_utf16(x, y);
        }
    }
    Ordering::Equal
}

/// `isNewer(candidate, current)`.
#[must_use]
pub fn is_newer(candidate: &str, current: &str) -> bool {
    let (Some(a), Some(b)) = (parse_version(candidate), parse_version(current)) else {
        return false;
    };
    for i in 0..3 {
        if a.release[i] != b.release[i] {
            return a.release[i] > b.release[i];
        }
    }
    match (&a.pre, &b.pre) {
        (None, pre) => pre.is_some(),
        (Some(_), None) => false,
        (Some(x), Some(y)) => compare_prerelease(x, y) == std::cmp::Ordering::Greater,
    }
}

/// `noticeFrom`.
#[must_use]
pub fn notice_from(cache: Option<&UpdateCache>, current: &str) -> Option<UpdateNotice> {
    let cache = cache?;
    if let Some((version, _)) = &cache.installed {
        if version == current {
            return None;
        }
        if is_newer(version, current) {
            return Some(UpdateNotice {
                latest: version.clone(),
                current: current.to_owned(),
                installed: true,
            });
        }
    }
    let latest = cache.latest.as_ref()?;
    if !is_newer(latest, current) {
        return None;
    }
    Some(UpdateNotice {
        latest: latest.clone(),
        current: current.to_owned(),
        installed: false,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn version_order_follows_the_engine() {
        assert!(is_newer("0.34.0", "0.33.0-rc.1"));
        assert!(is_newer("0.33.0", "0.33.0-rc.1"));
        assert!(!is_newer("0.33.0-rc.1", "0.33.0"));
        assert!(is_newer("0.33.0-rc.2", "0.33.0-rc.1"));
        assert!(is_newer("0.33.0-rc.10", "0.33.0-rc.9"));
        assert!(!is_newer("0.33.0", "0.33.0"));
        assert!(!is_newer("abc", "0.1.0"));
        assert!(is_newer("v1.0.0+build", "0.9.9"));
        assert!(is_newer("1.0.0-beta", "1.0.0-1"));
    }
}

File: packages/engine/src/cli/update-cache.ts
import { spawn } from 'node:child_process'
import { existsSync, mkdirSync, readFileSync, renameSync, writeFileSync } from 'node:fs'
import { homedir } from 'node:os'
import { basename, dirname, join, sep } from 'node:path'

/**
 * The update cache and the refresh claim (auto-update D1), split out of
 * update-check.ts so the boot stub can import it (rust-core 3.1).
 *
 * The boundary is "what the STUB must do after the native core has handled a
 * `statusline` or `status`": read the cache, decide whether a refresh is
 * due, claim the slot and spawn the detached `update-check --refresh` child.
 * `sofar-core` only ever READS the cache (rust-core O2 ruling: the binary
 * never spawns node), so without this the surfaces that trigger the daily
 * check today — the statusline on every prompt, `status` — would stop
 * triggering it the moment the binary owns them. Everything here is
 * dependency-free node builtins: the stub pays for every import on both the
 * hook and the human path.
 *
 * update-check.ts re-exports all of it; nothing else imports this module
 * except the stub.
 */

/** How long a completed check stays fresh. */
export const CHECK_TTL_MS = 24 * 60 * 60 * 1000

/** Set to any non-empty value to silence the check entirely. */
export const OPT_OUT_ENV = 'SOFAR_NO_UPDATE_CHECK'

export type Env = Record<string, string | undefined>

export interface UpdateCache {
  version: 1
  /** Last successfully resolved `latest` dist-tag; null when never resolved. */
  latest: string | null
  /** ISO timestamp of the last check ATTEMPT (claimed before the network call). */
  checked_at: string
  /** Set by an auto-install so a later surface can say wiring needs refreshing. */
  installed?: { version: string; at: string }
}

export function updateCachePath(env: Env = process.env): string {
  const base = nonEmpty(env.XDG_STATE_HOME) ?? join(homedir(), '.local', 'state')
  return join(base, 'sofar', 'update.json')
}

function nonEmpty(value: string | undefined): string | undefined {
  return value !== undefined && value.trim().length > 0 ? value : undefined
}

/**
 * The cache, or null when absent/unreadable/corrupt. NEVER throws: this is
 * read from the statusline and from init's tail, where a malformed cache file
 * must degrade to "no notice", never to a failed command.
 */
export function readUpdateCache(env: Env = process.env): UpdateCache | null {
  const path = updateCachePath(env)
  if (!existsSync(path)) return null
  try {
    const decoded = JSON.parse(readFileSync(path, 'utf8')) as Partial<UpdateCache> | null
    if (typeof decoded !== 'object' || decoded === null) return null
    if (typeof decoded.checked_at !== 'string' || decoded.checked_at.length === 0) return null
    const latest = typeof decoded.latest === 'string' && decoded.latest.length > 0 ? decoded.latest : null
    const installed =
      typeof decoded.installed === 'object' &&
      decoded.installed !== null &&
      typeof decoded.installed.version === 'string' &&
      typeof decoded.installed.at === 'string'
        ? decoded.installed
        : undefined
    return {
      version: 1,
      latest,
      checked_at: decoded.checked_at,
      ...(installed !== undefined ? { installed } : {}),
    }
  } catch {
    return null
  }
}

/** Write via temp + rename so a reader never sees a half-written file. */
export function writeUpdateCache(cache: UpdateCache, env: Env = process.env): void {
  const path = updateCachePath(env)
  try {
    mkdirSync(dirname(path), { recursive: true })
    const tmp = `${path}.${process.pid}.tmp`
    writeFileSync(tmp, `${JSON.stringify(cache, null, 2)}\n`, 'utf8')
    renameSync(tmp, path)
  } catch {
    // A cache we cannot persist costs a redundant check, never a broken command.
  }
}

// ---------------------------------------------------------------------------
// Where the binary lives — the only thing the refresh gate needs from upgrade.ts.
// ---------------------------------------------------------------------------

export type UpgradePlan =
  | { kind: 'global-npm'; prefix: string; selfPath: string }
  | { kind: 'not-global'; selfPath: string; reason: string }

/**
 * Decide whether — and where — this binary can self-upgrade, PURELY from its
 * own on-disk path. Deliberately does not consult `npm config get prefix`:
 * that value is exactly what lies when sofar was installed under a custom
 * prefix, which is the whole reason a naive global install misses the live
 * copy. npm's posix global layout is
 * `<prefix>/lib/node_modules/sofar.sh/…`, so the prefix is the
 * directory holding `lib`. Anything that is not that layout — a project-local
 * dependency, an npx cache, a source checkout, or a Windows global root — is
 * reported `not-global` with a reason, and the caller prints manual guidance
 * rather than guessing a prefix and installing into the wrong place.
 */
export function planUpgrade(selfPath: string): UpgradePlan {
  const segments = selfPath.split(sep)
  const nmIndex = segments.lastIndexOf('node_modules')
  if (nmIndex < 0) {
    return {
      kind: 'not-global',
      selfPath,
      reason: 'not running from an installed package (looks like a source checkout)',
    }
  }
  const beforeNodeModules = segments.slice(0, nmIndex).join(sep)
  // posix npm global: node_modules sits directly inside `lib`, and the prefix
  // is lib's parent. A local dep (`<project>/node_modules`) or an npx cache
  // (`…/_npx/<hash>/node_modules`) has some other parent and must not self-upgrade.
  if (basename(beforeNodeModules) !== 'lib') {
    return {
      kind: 'not-global',
      selfPath,
      reason: 'not a global npm install (local dependency, npx cache, or non-npm layout)',
    }
  }
  return { kind: 'global-npm', prefix: dirname(beforeNodeModules), selfPath }
}

// ---------------------------------------------------------------------------
// The refresh gate and the claim.
// ---------------------------------------------------------------------------

export interface RefreshContext {
  plan: UpgradePlan
  cache: UpdateCache | null
  now: number
  env: Env
}

/**
 * Should a refresh be spawned right now?
 *
 * Gated on global-npm because nothing else can act on the answer: a source
 * checkout (this repo, during development) has nothing to upgrade, a local
 * dependency is pinned by its own package.json, and an npx run already
 * resolves latest. Nagging any of them is noise with no button attached.
 */
export function shouldRefresh(ctx: RefreshContext): boolean {
  if (nonEmpty(ctx.env[OPT_OUT_ENV]) !== undefined) return false
  // Nobody is present to read a notice in CI or under a test runner, and both
  // spend a real network call plus a write to the USER's home state dir to
  // learn it. This is not hypothetical: sofar's own packaging test installs
  // the tarball into a temp prefix — a true global-npm layout — and drove a
  // live `npm view` out of a unit-test run before this line existed.
  if (nonEmpty(ctx.env.CI) !== undefined) return false
  if (nonEmpty(ctx.env.VITEST) !== undefined || ctx.env.NODE_ENV === 'test') return false
  if (ctx.plan.kind !== 'global-npm') return false
  if (ctx.cache === null) return true
  const checkedAt = Date.parse(ctx.cache.checked_at)
  if (Number.isNaN(checkedAt)) return true
  // A clock that moved backwards (or a cache from the future) reads as stale
  // rather than pinning the check off forever.
  return ctx.now - checkedAt >= CHECK_TTL_MS || checkedAt > ctx.now
}

/**
 * The bundle to re-launch for the refresh — always the sibling `cli.js`.
 *
 * NOT selfPath. The build splits the CLI into three bundles (boot → cli.js,
 * hot path → fast.js, everything else → full.js, speed-2 T1), and the surface
 * that most needs this check — the statusline — runs inside fast.js. Spawning
 * selfPath there would run fast.js as a script: it has no top-level entry, so
 * it would exit silently and the check would never once happen. cli.js is the
 * only file that routes a command.
 */
export function refreshEntry(selfPath: string): string {
  return join(dirname(selfPath), 'cli.js')
}

/** Detached, unref'd, output discarded — the parent exits without waiting. */
export function defaultSpawnRefresh(selfPath: string): void {
  try {
    const child = spawn(process.execPath, [refreshEntry(selfPath), 'update-check', '--refresh'], {
      detached: true,
      stdio: 'ignore',
    })
    child.unref()
  } catch {
    // No child, no check. Never fatal to the foreground command.
  }
}

export interface ClaimDeps {
  selfPath: string
  spawnRefresh?: (selfPath: string) => void
  now?: number
  env?: Env
}

/**
 * The claim every notice-rendering surface makes: when the cache has gone
 * stale, stamp `checked_at` and kick off a background refresh for NEXT time.
 * Returns the cache as it was BEFORE the claim — the notice is rendered from
 * that, so a claim never changes what this invocation prints.
 *
 * The parent claims the slot (stamps checked_at before spawning) because the
 * statusline renders on every prompt: without the claim, a stale cache would
 * spawn one `npm view` per keystroke-round until the first child finished.
 */
export function claimRefresh(deps: ClaimDeps): UpdateCache | null {
  const env = deps.env ?? process.env
  const cache = readUpdateCache(env)
  const now = deps.now ?? Date.now()
  if (shouldRefresh({ plan: planUpgrade(deps.selfPath), cache, now, env })) {
    writeUpdateCache(
      {
        version: 1,
        latest: cache?.latest ?? null,
        checked_at: new Date(now).toISOString(),
        ...(cache?.installed !== undefined ? { installed: cache.installed } : {}),
      },
      env,
    )
    ;(deps.spawnRefresh ?? defaultSpawnRefresh)(deps.selfPath)
  }
  return cache
}
